File: CS50/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// AMEX 15         STRT 34 OR 37
// MC   16         STRT 51, 52, 53, 54, 55
// VISA 13 OR 16   STRT 4

int luhnAlgo(long int card);
long power(int n, int p);

int main(void)
{
    long int card;
    int sum = 0;

    do
    {
        card = get_long("Card Number: ");
    }
    while (card < 1);

    int amex = card / power(10, 13);
    int master = card / power(10, 14);
    int visa1 = card / power(10, 12);
    int visa2 = card / power(10, 15);

    sum = luhnAlgo(card);

    // check type
    if (sum % 10 == 0)
    {
        if (amex == 34 || amex == 37)
        {
            printf("AMEX\n");
        }
        else if (master == 51 || master == 52 || master == 53 || master == 54 || master == 55)
        {
            printf("MASTERCARD\n");
        }
        else if (visa1 == 4 || visa2 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

// Luhn Algo
int luhnAlgo(long int card)
{
    int sum = 0;
    int digit_count = 0;
    while (card > 0)
    {
        int digit = card % 10;
        card /= 10;
        digit_count++;

        if (digit_count % 2 == 0)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }
        sum += digit;
    }
    return sum;
}

long power(int n, int p)
{
    long result = 1;
    for (int i = 0; i < p; i++)
    {
        result = result * n;
    }
    return result;
}

File: CS50/sentimental-credit/credit.py
# AMEX 15         STRT 34 OR 37
# MC   16         STRT 51, 52, 53, 54, 55
# VISA 13 OR 16   STRT 4

def main():
    # prompt user for input
    while True:
        try:
            card = int(input("Card Number: "))
            if card > 0:
                break
        except ValueError:
            continue

    # convert card into str to count the length
    str_card = str(card)
    length = len(str_card)

    amex = int(str_card[:2])
    master = int(str_card[:2])
    visa = int(str_card[0])

    _sum = luhn(str_card)

    # verify that change is 0 or not
    if _sum % 10 == 0:
        # AMEX verification
        if length == 15 and amex in [34, 37]:
            print("AMEX")
        # MASTERCARD verification
        elif length == 16 and master in range(51, 56):
            print("MASTERCARD")
        # AMEX verification
        elif length in [13, 16] and visa == 4:
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
    exit(0)

# luhn function to calculate total coins


def luhn(card):
    _sum = 0
    rev_card = card[::-1]
    for i, digit_char in enumerate(rev_card):
        digit = int(digit_char)
        if i % 2 == 1:
            digit *= 2
            if digit > 9:
                digit -= 9
        _sum += digit
    return _sum


if __name__ == "__main__":
    main()
